File: src/hack_forked/utils/cgroup/cGroup.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Core_kernel
open Result.Monad_infix

let spf = Printf.sprintf

(* Little helper module to help memoize things. Probably could be pulled out into its own module
 * at some point *)
module Memoize : sig
  val forever : f:(unit -> 'a Lwt.t) -> unit -> 'a Lwt.t

  val until : seconds:float -> f:(unit -> 'a Lwt.t) -> unit -> 'a Lwt.t
end = struct
  let forever ~f =
    let memoized_result = ref None in
    fun () ->
      match !memoized_result with
      | None ->
        let%lwt result = f () in
        memoized_result := Some result;
        Lwt.return result
      | Some result -> Lwt.return result

  let until ~seconds ~f =
    let memoized_result = ref None in
    let fetch () =
      let%lwt result = f () in
      memoized_result := Some (Unix.gettimeofday () +. seconds, result);
      Lwt.return result
    in
    fun () ->
      match !memoized_result with
      | Some (good_until, result) when Unix.gettimeofday () < good_until -> Lwt.return result
      | _ -> fetch ()
end

(* Like Sys_utils.cat but with lwt. Probably also could be moved to its own module *)
let cat file =
  (* Some low level system operations cannot be canceled and will just ignore cancellation. If we
   * wrap this code in `Lwt.protected` we get a thread that is cancellable while the underlying
   * operations proceed as normal. *)
  Lwt.protected
    (try%lwt
       let%lwt ic =
         Lwt_io.open_file
           ~flags:[Unix.O_RDONLY; Unix.O_NONBLOCK]
           ~mode:Lwt_io.Input
           ~perm:0o666
           file
       in
       let%lwt contents = Lwt_io.read ic in
       let%lwt () = Lwt_io.close ic in
       Lwt.return_ok contents
     with
    | Unix.Unix_error (Unix.ENOENT, _, _) -> Lwt.return_error "File doesn't exist"
    | Unix.Unix_error (Unix.EBADF, _, _) -> Lwt.return_error "Lost fd for the file")

(* I've never seen cgroup mounted elsewhere, so it's probably fine to hardcode this for now *)
let cgroup_dir = "/sys/fs/cgroup"

let assert_is_using_cgroup_v2 =
  Memoize.forever ~f:(fun () ->
      if%lwt Lwt_unix.file_exists cgroup_dir then
        (* /sys/fs/cgroup/memory exists for cgroup v1 but not v2. It's an easy way to tell the
         * difference between versions *)
          if%lwt Lwt_unix.file_exists (spf "%s/memory" cgroup_dir) then
            Lwt.return_error (spf "cgroup v1 is mounted at %s. We need v2" cgroup_dir)
          else
            Lwt.return_ok ()
      else
        Lwt.return_error (spf "%s doesn't exist" cgroup_dir))

(* I don't really expect us to switch cgroups often, but let's only cache for 5 seconds *)
let get_cgroup_name =
  Memoize.until ~seconds:5.0 ~f:(fun () ->
      ProcFS.first_cgroup_for_pid (Unix.getpid ()) |> Lwt.return)

type stats = {
  total: int;
  (* The total physical memory for the cgroup *)
  total_swap: int;
  (* The total amount of anonymous memory paged out to swap *)

  (* anon, file, and shmem are disjoint. If you add in the memory that the kernel uses, they should
   * sum roughly to `total` *)
  anon: int;
  (* The amount of physical anonymous memory not used for shared memory *)
  shmem: int;
  (* The amount of physical anonymous memory being used as shared memory *)
  file: int; (* The amount of physical memory which is not anonymous *)
}

(* Some cgroup files contain only a single integer *)
let read_single_number_file path =
  let%lwt contents_result = cat path in
  Lwt.return
    ( contents_result >>= fun contents ->
      try Ok (contents |> String.strip |> int_of_string)
      with Failure _ -> Error "Failed to parse memory.current" )

let parse_stat stat_contents =
  let stats =
    String.split stat_contents ~on:'\n'
    |> List.fold_left ~init:SMap.empty ~f:(fun stats line ->
           match String.split line ~on:' ' with
           | [key; raw_stat] ->
             int_of_string_opt raw_stat
             |> Base.Option.value_map ~default:stats ~f:(fun stat -> SMap.add key stat stats)
           | _ -> stats)
  in
  let get key =
    match SMap.find_opt key stats with
    | Some stat -> Ok stat
    | None -> Error (spf "Failed to find %S in memory.stat" key)
  in
  get "anon" >>= fun anon ->
  get "file" >>= fun file ->
  get "shmem" >>| fun shmem ->
  (* In `memory.stat` the `file` stat includes `shmem` *)
  (anon, file - shmem, shmem)

let get_stats_for_cgroup (cgroup_name : string) : (stats, string) result Lwt.t =
  (* cgroup_name starts with a /, like /my_cgroup *)
  let dir = spf "%s%s" cgroup_dir cgroup_name in
  let%lwt total_result = read_single_number_file (Filename.concat dir "memory.current")
  and total_swap_result = read_single_number_file (Filename.concat dir "memory.swap.current")
  and stat_contents_result = cat (Filename.concat dir "memory.stat") in
  Lwt.return
    ( total_result >>= fun total ->
      total_swap_result >>= fun total_swap ->
      stat_contents_result >>= fun stat_contents ->
      parse_stat stat_contents >>= fun (anon, file, shmem) ->
      Ok { total; total_swap; anon; file; shmem } )

(* Like Result's >>= but for when you're dealing with result threads *)
let ( >>% ) (type a b c) (thread : (a, b) result Lwt.t) (f : a -> (c, b) result Lwt.t) :
    (c, b) result Lwt.t =
  match%lwt thread with
  | Ok value -> f value
  | Error e -> Lwt.return_error e

let get_stats () = assert_is_using_cgroup_v2 () >>% get_cgroup_name >>% get_stats_for_cgroup

File: rust_port/crates/flow_server/src/cgroup.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::sync::Mutex;
use std::sync::OnceLock;
use std::time::SystemTime;
use std::time::UNIX_EPOCH;

pub struct Stats {
    pub total: i64,
    pub total_swap: i64,
    pub anon: i64,
    pub shmem: i64,
    pub file: i64,
}

const CGROUP_DIR: &str = "/sys/fs/cgroup";

fn assert_is_using_cgroup_v2() -> Result<(), String> {
    static RESULT: OnceLock<Result<(), String>> = OnceLock::new();
    RESULT
        .get_or_init(|| {
            if std::path::Path::new(CGROUP_DIR).exists() {
                let memory_path = format!("{}/memory", CGROUP_DIR);
                if std::path::Path::new(&memory_path).exists() {
                    Err(format!(
                        "cgroup v1 is mounted at {}. We need v2",
                        CGROUP_DIR
                    ))
                } else {
                    Ok(())
                }
            } else {
                Err(format!("{} doesn't exist", CGROUP_DIR))
            }
        })
        .clone()
}

fn get_cgroup_name() -> Result<String, String> {
    static CACHE: Mutex<Option<(f64, Result<String, String>)>> = Mutex::new(None);
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();
    let mut cache = CACHE.lock().unwrap();
    if let Some((good_until, ref result)) = *cache {
        if now < good_until {
            return result.clone();
        }
    }
    let result = flow_procfs::first_cgroup_for_pid(std::process::id());
    *cache = Some((now + 5.0, result.clone()));
    result
}

fn cat(file: &str) -> Result<String, String> {
    std::fs::read_to_string(file).map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            "File doesn't exist".to_string()
        } else {
            format!("{}", e)
        }
    })
}

fn read_single_number_file(path: &str) -> Result<i64, String> {
    let contents = cat(path)?;
    contents
        .trim()
        .parse::<i64>()
        .map_err(|_| "Failed to parse memory.current".to_string())
}

fn parse_stat(stat_contents: &str) -> Result<(i64, i64, i64), String> {
    let stats: BTreeMap<&str, i64> =
        stat_contents
            .split('\n')
            .fold(BTreeMap::new(), |mut stats, line| {
                let parts: Vec<&str> = line.split(' ').collect();
                if parts.len() == 2 {
                    if let Ok(stat) = parts[1].parse::<i64>() {
                        stats.insert(parts[0], stat);
                    }
                }
                stats
            });
    let get = |key: &str| -> Result<i64, String> {
        stats
            .get(key)
            .copied()
            .ok_or_else(|| format!("Failed to find {:?} in memory.stat", key))
    };
    let anon = get("anon")?;
    let file = get("file")?;
    let shmem = get("shmem")?;
    Ok((anon, file - shmem, shmem))
}

// cgroup_name starts with a /, like /my_cgroup
fn get_stats_for_cgroup(cgroup_name: &str) -> Result<Stats, String> {
    let dir = format!("{}{}", CGROUP_DIR, cgroup_name);
    let total = read_single_number_file(&format!("{}/memory.current", dir))?;
    let total_swap = read_single_number_file(&format!("{}/memory.swap.current", dir))?;
    let stat_contents = cat(&format!("{}/memory.stat", dir))?;
    let (anon, file, shmem) = parse_stat(&stat_contents)?;
    Ok(Stats {
        total,
        total_swap,
        anon,
        file,
        shmem,
    })
}

pub fn get_stats() -> Result<Stats, String> {
    assert_is_using_cgroup_v2()?;
    let cgroup_name = get_cgroup_name()?;
    get_stats_for_cgroup(&cgroup_name)
}
